### PR TITLE
[Order] Order properties and constants with respect of complexity and position

### PR DESCRIPTION
**Pull request by dobryy**
Closes https://github.com/rectorphp/rector/issues/3807

### CODE REVIEW

**Comment by dobryy** [comment]
@TomasVotruba This is not the final version but I'd like to share the progress. Some decoupling is required.

To make it work correctly I had to sort by visibility, complexity, and name. There is still one problem with Traits. I created a test fixture for it and will look into it tomorrow.

**Comment by TomasVotruba** [comment]
There are 3 fixes by Rector CI now. There should be just 2:

- 1 for Rector fixes
- and 1 for coding standard fixing the changes

image

That probably means Rector is needed to run twice to work, which is not good. I should be just once.
Could you fix that before review?

**Comment by dobryy** [comment]
> There are 3 fixes by Rector CI now. There should be just 2:
> 
> * 1 for Rector fixes
> * and 1 for coding standard fixing the changes
> 
> image
> 
> That probably means Rector is needed to run twice to work, which is not good. I should be just once.
> Could you fix that before review?

Everything is fine, it runs only two times. It looks like three times because I squashed two previous fixups into one and then it created another two after the rebase. I'll reset the branch to only my changes to make sure it runs fine.

**Comment by dobryy** [comment]
> That probably means Rector is needed to run twice to work, which is not good. I should be just once.

Is this relevant to all rules? It seems like for ordering rules (e.g. OrderPrivateMethodsByUseRector) it is very possible that Rector will need to run couple of times.

**Comment by TomasVotruba** [comment]
> Everything is fine, it runs only two times. It looks like three times because I squashed two previous fixups into one and then it created another two after the rebase. I'll reset the branch to only my changes to make sure it runs fine.

Ah, I see. Thanks for resetting!

**Comment by TomasVotruba** [comment]
> Is this relevant to all rules? It seems like for ordering rules (e.g. OrderPrivateMethodsByUseRector) it is very possible that Rector will need to run couple of times.

It's bad practise and should be avoided if possible. All Rector rules should be deterministics. If it's not the case, it's mostly likely lack of effort :D

**Comment by dobryy** [comment]
> > Is this relevant to all rules? It seems like for ordering rules (e.g. OrderPrivateMethodsByUseRector) it is very possible that Rector will need to run couple of times.
> 
> It's bad practise and should be avoided if possible. All Rector rules should be deterministic:
> 
> * A → B  heavy_check_mark
> * ~A → B → C~ no_entry
> 
> If it's not the case, it's mostly likely lack of effort :D

Lack of effort? It's not about us! :D

File: rules/order/src/Rector/Class_/OrderPropertyByComplexityRector.php [hunk 1]
```diff
@@ -100,50 +102,53 @@ public function refactor(Node $node): ?Node
...
 
-            foreach ($propertyByPosition as $position => $property) {
+            foreach ($resolvePropertyByPosition as $position => $property) {
                 /** @var string $propertyName */
                 $propertyName = $this->getName($property);
 
                 $propertyPositionByName[$position] = $propertyName;
-                $propertyNameToRank[$propertyName] = $this->propertyRanker->rank($property);
+                $propertyNameToRank[$propertyName]['rank'] = $this->propertyRanker->rank($property);
+                $propertyNameToRank[$propertyName]['position'] = $position;
```
Comment (TomasVotruba) [inline rules/order/src/Rector/Class_/OrderPropertyByComplexityRector.php:115]: I see. This is the spot that fixes it all? :)

File: rules/order/src/Rector/Class_/OrderPropertyByComplexityRector.php [hunk 2]
```diff
@@ -100,50 +102,53 @@ public function refactor(Node $node): ?Node
...
     /**
-     * @return Property[][]
+     * @return Property[][]|ClassConst[][]
      */
     private function resolvePropertyByVisibilityByPosition(Class_ $class): array
     {
         $propertyByVisibilityByPosition = [];
         foreach ($class->stmts as $position => $classStmt) {
-            if (! $classStmt instanceof Property) {
+            if (! $classStmt instanceof Property && ! $classStmt instanceof ClassConst) {
```
Comment (TomasVotruba) [inline rules/order/src/Rector/Class_/OrderPropertyByComplexityRector.php:151]: I see `OrderPropertyByComplexityRector` but here it orderes `ClassConst` as well? That's confusing.

It should be 2 separated rules or ClassConst skipped here.

File: rules/order/src/Rector/Class_/OrderPropertyByComplexityRector.php [hunk 3]
```diff
@@ -153,4 +158,13 @@ private function resolvePropertyByVisibilityByPosition(Class_ $class): array
...
     }
+
+    private function getSortedPropertiesByRankAndPosition(array $propertyNameToRank): array
+    {
+        $rank = array_column($propertyNameToRank, 'rank');
+        $position = array_column($propertyNameToRank, 'position');
+        array_multisort($rank, SORT_ASC, $position, SORT_ASC, $propertyNameToRank);
+
+        return array_keys($propertyNameToRank);
+    }
```
Comment (TomasVotruba) [inline rules/order/src/Rector/Class_/OrderPropertyByComplexityRector.php:169]: This is very hard read to me. What should it do?

Wouldn't battleship do it?

See the last code snippet here: https://blog.martinhujer.cz/clever-way-to-sort-php-arrays-by-multiple-values/

File: rules/order/src/StmtOrder.php [hunk 4]
```diff
@@ -44,7 +44,7 @@ public function reorderClassStmtsByOldToNewKeys(Class_ $class, array $oldToNewKe
             // reorder here
             $newKey = $oldToNewKeys[$key];
 
-            $reorderedStmts[$newKey] = $stmt;
+            $reorderedStmts[$key] = $class->stmts[$newKey];
```
Comment (TomasVotruba) [inline rules/order/src/StmtOrder.php:47]: This looks like a bug fix. Why this change?

**Comment by TomasVotruba** [comment]
I've added few comments
